README.md
Disclaimer
==========

*This project is currently in development. Do not use it on production
environments!*

About
=====

`pgapi` tries to be a simple but useful REST API for Debian based
PostgreSQL Systems.  It offers a way to create, delete and control
PostgreSQL clusters via HTTP requests.

pgapi is built to be extensible. This means all API endpoints are
implemented using their own modules. At the moment pgapi ships with 
two modules.

  1. **cluster**: a /cluster endpoint to create, drop,
     start, stop and configure PostgreSQL Clusters.
  2. **system**: a /system endpoint that provides information
     about the system it is running on.

Requirements
============

The user executing the api needs privileges to run commands as user
postgres using sudo.  The user postgres needs sudo privileges to start
and stop a cluster via pg_ctlcluster as user root.

A example sudo configuration could look like this:

```
<myuser>	ALL=(postgres:ALL) NOPASSWD:ALL
postgres	ALL=(root) NOPASSWD: /usr/bin/pg_ctlcluster *
```

Getting Started
===============

1. checkout this git repository:

   ```
   git clone https://github.com/credativ/pgapi
   ```

2. change to the new directory:

   ```
   cd pgapi
   ```

3. start the standalone service:

   ```
   pgapi/api.py --config-file conf/pgapi.conf
   ```

4. get a list of all clusters:

   ```
   curl -X GET http://127.0.0.1:15432/cluster/
   ```

Installation
============

Using setuptools
----------------

`python setuptools` is within the installation process.

```
python3 setup.py build
python3 setup.py install
```

Examples
========

Cluster module:
---------------

* get a list of all clusters:
  `curl -X GET http://127.0.0.1:15432/cluster/`

* create a new cluster with version 9.6 and name foobar:
  `curl -X POST http://127.0.0.1:15432/cluster/9.6/foobar`

* change port of the 9.6/foobar cluster:
  ```
  curl -i -X PATCH \
       -H "Content-Type: application/json" \
       -d '{"config": {"port": 8432}}' \
       http://127.0.0.1:15432/cluster/9.6/foobar
  ```

* restart the 9.6/foobar cluster:
  ```
  curl -i -X PATCH \
       -H "Content-Type: application/json" \
	   -d '{"state": "restart"}' \
       http://127.0.0.1:15432/cluster/9.6/foobar
  ```

* change port once more and restart at once:
  ```
  curl -i -X PATCH \
       -H "Content-Type: application/json" \
       -d '{"config": {"port": 9432}, "state": "restart"}' \
       http://127.0.0.1:15432/cluster/9.6/foobar
  ```

* delete the cluster 9.6/foobar:
  `curl -X DELETE http://127.0.0.1:15432/cluster/9.6/foobar`

System module:
---------------

* get all system information available via the system module:
  ```
  curl -X GET http://127.0.0.1:15432/system/
  ```

* get a specific system section information (see previous example for
  available sections):
  ```
  curl -X GET http://127.0.0.1:15432/system/hostname/
  ```
  ```
  curl -X GET http://127.0.0.1:15432/system/installed_pg_versions/
  ```


Python Dependencies
===================

See requirements.txt for a full list of dependencies.

Other Dependencies
==================

Beside the above listed python dependencies, it's expected that
postgresql-common is available on the system.  pgapi makes heavy use
of postgresql-common and some of the tools that are shipped with it.

Licence
=======

This project uses the MIT Licence.

Debugging
=========

If something doesn't work as expected, try to start pgapi in
debug mode:
```
pgapi/api.py --config-file conf/pgapi.conf --debug
```

Tests
=====
Testing requires the packages pytest and pytest-cov 
( apt: python3-pytest python3-pytest-cov ) to be installed.

To run all unit tests execute `runtests.sh`

```
$ ./runtests.sh
```


debian/changelog
pgapi (0.0.5) UNRELEASED; urgency=medium

  * 

 -- Christoph Berg <myon@debian.org>  Wed, 28 Aug 2019 15:13:37 +0200

pgapi (0.0.4) unstable; urgency=medium

  [ Julian Schauder ]
  * Added Arguments to Clustercreation
  * Initial implementation for pg_controldata
  * Added dependencies to tests
  * Bumped default tests-pgversion from 9.5 to 9.6. this is default in debian stable(stretch)
  * Minor fixes to the README

  [ Adrian Vondendriesch ]
  * Add "load" section to systemCommands.
  * Refactor get_installed_postgresql_versions.
  * Respect listen_address parameter.

 -- Adrian Vondendriesch <adrian.vondendriesch@credativ.de>  Tue, 09 Oct 2018 21:33:39 +0200

pgapi (0.0.3) unstable; urgency=medium

  [ Adrian Vondendriesch ]
  * Implement a new HTTP endpoint "/system".

  [ Julian Schauder ]
  * Added wsgi as apache dependency

  [ Adrian Vondendriesch ]
  * Add additional system information to /system endpoint.
  * Add psutil to requirements.
  * Add some comments about the behavior of systemAPI.
  * Update README to describe the new system endpoint.

 -- Adrian Vondendriesch <adrian.vondendriesch@credativ.de>  Wed, 19 Sep 2018 14:19:56 +0200

pgapi (0.0.2) unstable; urgency=medium

  [ Julian Schauder ]
  * Added some hardening and errormessages
  * Increased codecoverage for tests

  [ Adrian Vondendriesch ]
  * Restructure setup.py:
  * Change plugin import method.
  * Change default settings that define user handling.
  * Add WSGI sample files.
  * Make pgapi available as console script.
  * Add postgresql-common (>= 182) to debian depends.
  * Introduce a new debian package pgapi-apache2.
  * Turned debug off by default, move default pgapi.conf.
  * Install default pgapi.conf under /etc/pgapi.
  * Remove tabs from pgapi.conf since it's yaml.
  * Correct config handling.

  [ Julian Schauder ]
  * Fixed shutdown-mode for restart and stop actions

  [ Adrian Vondendriesch ]
  * Fix bypass_systemd logic.

 -- Adrian Vondendriesch <adrian.vondendriesch@credativ.de>  Tue, 18 Sep 2018 18:18:43 +0200

pgapi (0.0.1) unstable; urgency=low

  * Initial Debian packaging.

 -- Adrian Vondendriesch <adrian.vondendriesch@credativ.de>  Fri, 23 Mar 2018 18:05:01 +0100


pgapi/helper.py
import logging
import os
import yaml
import re

class Config(object): 
    __config = {
        "defaults": {
            "name": "PGAPI",
            "port": 15432,
            "listen_address": "127.0.0.1",
            "debug": False,
            "sudo_user": "postgres",
            "use_sudo": False,
            "config_file": "/etc/pgapi/pgapi.conf",
            "ssl_mode": "off", # off or adhoc
            "bypass_systemd": True,
            "encoding": 'UTF-8'
        },
        "file": {},
        "arguments": {}
    }

    __namespaces = ["arguments", "file", "defaults"]

    __instance = None

    @classmethod
    def getInstance(cls):
        """Get a Config object instance.

        If not already created, it will create a new Config object.
        
        Returns (Config):
           config -- Config object
        """
        if not cls.__instance:
            cls.__instance = Config()
        return cls.__instance
        
    def getSettingAndNamespace(self, setting):
        """Returns the value for a given setting and provides the
        namespace it comes from.

        Function Arguments:
           setting (string) -- setting for which the value will be returned
        
        Returns: tuple (value, namespace)
           value (any)        -- value for setting, if setting is found in any namespace
                              -- None otherwise
           namespace (string) -- namespace, if setting is found in any namespace
                              -- None otherwise
        """
        logging.debug("search for setting: \"%s\"", setting)
        for namespace in self.__namespaces:
            if setting in self.__config[namespace]:
                return (self.__config[namespace][setting], namespace)
        return (None, None)

    def getSetting(self, setting):
        """Returns the value for a given setting.
        """
        (value, _) = self.getSettingAndNamespace(setting)
        return value

    def setSetting(self, namespace, setting, value):
        """Update a setting for a give namespace.
        Valid namespaces are "defaults", "file" and "arguments"

        Function Arguments:
            namespace (string) -- namespace for setting
            setting (string)   -- setting to set
            value (any)        -- value to save

        Returns (bool):
            True -- always

        Throws:
            KeyError -- if namespace is invalid
        """
        logging.debug("update \"%s\" to \"%s\" (namespace: \"%s\")", setting, value, namespace)
        if not namespace in self.__namespaces:
            raise KeyError("\"%s\" is not a valid namespace", namespace)

        self.__config[namespace][setting] = value

        return True

    def loadConfigFile(self, filename):
        """Load a given yaml config into "file" namespace.

        Returns (bool):
            True  -- if the file could be loaded and
            False -- if the file could not be loaded."""
        try:
            with open(filename, 'r') as f:
                self.__config["file"] = yaml.load(f)
        except:
            return False

        return True

def check_regex(pattern, value):
    """Checks a given value against a given regex pattern.

    Function Argmuents:
       pattern (string) -- regex pattern
       value (string)   -- value to compair against the pattern
    
    Returns (bool):
       True  -- if the value matches the provided pattern
       False -- otherwise
    """
    found = re.search(pattern, value)
    if not found:
        return False

    return True
    

def param_is_safe(param):
    """
    Checks if a postgresql parameter is "safe".
      We only allow the following characters:
      * [a-z][A-Z]
      * "_"

    Function Arguments:
       param (string) -- parameter to check

    Returns (bool):
       True  -- if param seems to be valid (regarding the defined regex)
       False -- otherwise
    """
    pattern = r"^[\w_]*.*[a-zA-Z]+[\w]*$"
    return check_regex(pattern, param)

def valid_cluster_name(name):
    """Checks if a cluster name is valid.
    
    Function Arguments:
       name (string) -- name of the cluster
    
    Returns (bool):
       True  -- if the cluster name is valid (regarding the defined regex)
       False -- otherwise
    """
    pattern = r"^[a-zA-Z]+[\w\-_]*$"
    return check_regex(pattern, name)

def valid_cluster_version(version):
    """Checks if a given cluster version is valid.
    
    Function Arguments:
       name (string) -- version of the cluster
    
    Returns (bool):
       True  -- if the cluster version is valid
       False -- otherwise
    """
    # We want to match eg. 10, 9.6 but not 9.6.1 or 10.1
    pattern = r"^(\d(\.\d)?|\d{2})$"
    return check_regex(pattern, version)

def get_installed_postgresql_versions():
    """Returns a list of installed PostgreSQL versions.

       Returns (dict):
         key   (string)  -- PostgreSQL version (e.g. 9.6 or 10)
         value (dict)    -- dict containing additional information about this version
                            (e.g. binary path)
    """
    versions = {}
    install_dir = "/usr/lib/postgresql/"

    # Return a empty version list if no postgresql verion, if the
    # install_dir does not exist.
    if not os.path.isdir(install_dir):
        return versions
    
    subdirs = os.listdir(install_dir)

    for version_dir in subdirs:
        initdb_path = os.path.join(*[install_dir, version_dir, "bin", "initdb"])
        if os.path.isfile(initdb_path):
            versions[version_dir] = {}
            versions[version_dir]['binary_path'] = os.path.join(*[install_dir, version_dir, "bin"])

    return versions
